src/Set.js

```diff
@@ -82,19 +82,23 @@ export class Set extends SetCollection {
   // @pragma Composition
 
   map(mapper, context) {
-    const removes = [];
-    const adds = [];
-    this.forEach((value) => {
-      const mapped = mapper.call(context, value, value, this);
-      if (mapped !== value) {
-        removes.push(value);
-        adds.push(mapped);
-      }
-    });
-    return this.withMutations((set) => {
-      removes.forEach((value) => set.remove(value));
-      adds.forEach((value) => set.add(value));
-    });
+    // keep track if the set is altered by the map function
+    let didChanges = false;
+
+    const newMap = updateSet(
```

Comment:
I reused `updateSet` here as it was the base implementation (see [here](https://github.com/immutable-js/immutable-js/pull/1606/files#diff-f6aa3e181d81c115fe15d0c5d744a41f4d9c1f95f811e05137894ea215f3370d) for the original implementation)